pkg/service/backup/worker_deduplicate.go

```diff
@@ -0,0 +1,219 @@
+	remoteSSTables map[string]struct{}, ssTablesGroupByID map[string][]string,
+) (deduplicated int64, err error) {
+	// Per every SSTable files group, check if the name uses UUID to identify the SSTable generation.
+	// If the above is true, then check if files exists in the remote storage and remove it locally if exists in rmeote.
+	for id, ssTableContent := range ssTablesGroupByID {
+		// Check if id is an integer (no UUID)
+		if _, err := strconv.Atoi(id); err == nil {
+			continue
+		}
+		// Check if SSTable is available on remote already
+		if _, ok := remoteSSTables[id]; !ok {
+			continue
+		}
+		// Remove duplicated SSTable from local snapshot
+		for _, file := range ssTableContent {
+			// remove if exists in remote
+			localSSTableFileNameWithPath := path.Join(d.Path, file)
+			w.Logger.Debug(ctx, "Removing local snapshot file (deduplication based on generation UUID)",
+				"host", h.IP, "file", localSSTableFileNameWithPath)
+			if err := w.Client.RcloneDeleteFile(ctx, h.IP, localSSTableFileNameWithPath); err != nil {
+				return deduplicated, errors.Wrapf(err, "cannot delete local snapshot's SSTable file %s", localSSTableFileNameWithPath)
+			}
+			idx := slices.IndexFunc(d.Progress.files, func(info fileInfo) bool {
+				return info.Name == file
+			})
+			if idx == -1 {
+				return deduplicated, fmt.Errorf("no file %s in indexed files", file)
+			}
+			deduplicated += d.Progress.files[idx].Size
+		}
+	}
+
+	return deduplicated, nil
+}
+
+func (w *worker) basedOnCrc32Content(ctx context.Context, d snapshotDir, h hostInfo, dataDst string,
+	remoteSSTables map[string]struct{}, ssTablesGroupByID map[string][]string,
+) (deduplicated int64, err error) {
+	// Reference to SSTables 3.0 Data File Format
+	// https://opensource.docs.scylladb.com/stable/architecture/sstable/sstable3/sstables-3-data-file-format.html
+
+	// Per every SSTable files group, compare local <ID>-Digest.crc32 content
+	// to the remote <ID>-Digest.crc32 content.
+	// The same content implies that SSTable can be deduplicated and removed from local directory.
+	for id, ssTableContent := range ssTablesGroupByID {
+		crc32Idx := slices.IndexFunc(ssTableContent, func(s string) bool {
+			return strings.HasSuffix(s, "Digest.crc32")
+		})
+		if crc32Idx == -1 {
+			continue
+		}
+		if _, ok := remoteSSTables[id]; !ok {
+			continue
+		}
+
+		remoteCRC32FileNameWithPath := path.Join(dataDst, ssTableContent[crc32Idx])
+		remoteCRC32, err := w.Client.RcloneCat(ctx, h.IP, remoteCRC32FileNameWithPath)
+		if err != nil {
+			if scyllaclient.StatusCodeOf(err) == 404 {
+				continue
+			}
+			return deduplicated, errors.Wrapf(err, "cannot get content of remote CRC32 %s", remoteCRC32FileNameWithPath)
+		}
+
+		localCRC32FileNameWithPath := path.Join(d.Path, ssTableContent[crc32Idx])
+		localCRC32, err := w.Client.RcloneCat(ctx, h.IP, localCRC32FileNameWithPath)
+		if err != nil {
+			if scyllaclient.StatusCodeOf(err) == 404 {
+				continue
+			}
```

Comment:
But how can this file be non-existant if we checked that it exists in the slice above?

Reply:
yeah, good point. I'm removing this check for remote crc32

Reply:
I think that this check should be removed for local crc 32, since we check if this file exists a few lines above:
```go
		crc32Idx := slices.IndexFunc(ssTableContent, func(s string) bool {
			return strings.HasSuffix(s, "Digest.crc32")
		})
		if crc32Idx == -1 {
			continue
		}
		if _, ok := remoteSSTables[id]; !ok {
			continue
		}

		remoteCRC32FileNameWithPath := path.Join(dataDst, ssTableContent[crc32Idx])
		remoteCRC32, err := w.Client.RcloneCat(ctx, h.IP, remoteCRC32FileNameWithPath)
		if err != nil {
			return deduplicated, errors.Wrapf(err, "cannot get content of remote CRC32 %s", remoteCRC32FileNameWithPath)
		}

		localCRC32FileNameWithPath := path.Join(d.Path, ssTableContent[crc32Idx])
		localCRC32, err := w.Client.RcloneCat(ctx, h.IP, localCRC32FileNameWithPath)
		if err != nil {
			if scyllaclient.StatusCodeOf(err) == 404 {
				continue
			}
```

Reply:
Aii, you are right. I removed it from remote, but should stay in remote.
Fixed.